### ISSUE TITLE
Illegal instructions error.

### ISSUE THREAD

**Issue by caonetto** [isovic/racon#47]
Hi, this is the first time I use Racon, and keep on getting these error when trying to run the command.

```
ubuntu@biolinux:/mnt/Federico/TD_4$ racon '/mnt/Federico/TD_2/trimmedpass.fastq' '/mnt/Federico/TD_4/mappings.paf' '/mnt/Federico/TD_4/smartdenovoTD1.fasta' racon1.fasta 
Illegal instructions.
```

I build racon using bioconda, so I guess the installation is ok?

Any ideas?

Thanks.

**Comment by rvaser** [comment]
Hello,
sorry for the late response! The error indicates that your processor does not support the same SIMD instruction as the ones in the bioconda package so you should recompile Racon on your machine.

Best regards,
Robert

**Comment by ebioman** [comment]
Hi
I have a similar problem when trying to provide RPM packages for different kind of machines (e.g. Intel vs AMD). Do you have a flag to drop some of these features and if yes, what would you guess would be the speed impact ?

**Comment by rvaser** [comment]
Racon supports SSE4.1+ and AVX2 which are automatically detected at compile time (you have to compile the source on each machine). If they are not supported, vectorization is not used and the program slows down between 2-4 times.

Best regards,
Robert

**Comment by ebioman** [comment]
Thanks, will see whether we can tackle this differently as we have one repository for different architectures

**Comment by bgruening** [comment]
Can we force compilation with SSE2 somehow?

**Comment by mmiladi** [comment]
Would you expect this can make a generic binary compatible over different architectures?
`cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=${PREFIX} -Dracon_build_wrapper=ON -DCMAKE_CXX_FLAGS="-mno-avx2 -mno-sse4.1 -mno-sse4.2 " ..`

**Comment by rvaser** [comment]
Hi guys,
I have hard coded ifdefs in SIMD part of the code which only compile when AVX2 or SSE4.1 are enabled (if both are missing, SISD version of the algorithm is used instead). Fortunately, it seems sse2 instruction set contains all the intrinsics that I use (I think some older code needed one or two instructions from SSE4.1). I will update everything soon :)

Best regards,
Robert

**Comment by mmiladi** [comment]
Hi Robert,
Sounds really good :)
Thanks,
Milad

**Comment by rvaser** [comment]
I am sorry but `_mm_max_epi32` is supported only in SSE4.1 and upwards so I can't downgrade to SSE2 (this intrinsic slipped out of my sight when I was checking compatibility).

Best regards,
Robert

**Comment by mmiladi** [comment]
I think most of incompatibility is due to `avx` support.  `sse4a` should be fine in more of the cases.

**Comment by mmiladi** [comment]
For example anaconda built binary differs in avx instructions, comparing with the locally built one on my laptop with `avx` support and not `avx2`:

```
$ diff local_set.txt anaconda_set.txt 
< vextractf128
< vinsertf128
< vmovaps
< vmovups
> vpalignr
< vpshufb
< vpshufd
< vpslldq

```

**Comment by rvaser** [comment]
Does anaconda provide a recipe to build the binary or does it comes with a precompiled one? I have never used it before nor did I release the package.

**Comment by mmiladi** [comment]
It's pretty simple, one needs a build recipe and meta information about dependencies and so on:
Please check here:
https://github.com/bioconda/bioconda-recipes/blob/master/recipes/racon

The build.sh is the compile recipe. Once a PR is merged into bioconda-recipes master branch, Anacoda automatically builds and store new binaries here:

https://anaconda.org/bioconda/racon/files

**Comment by rvaser** [comment]
If you disable sse4.1 you get a speed degradation of ~2-4 times. You should just disable avx2 as it is marginally faster than sse4.1.

**Comment by mmiladi** [comment]
Update: Using the vanilla installation the binary (immediately?) crashes:
```
racon sample_reads.fasta sample_overlaps.sam sample_layout.fasta
Illegal instruction
```

Built with  cmake option `-DCMAKE_CXX_FLAGS="-mno-avx2 -mno-sse4.1 -mno-sse4.2 "`, it crashed somewhere further :-D :+1: 
```
racon sample_reads.fasta sample_overlaps.sam sample_layout.fasta 
[racon::Polisher::initialize] loaded target sequences
[racon::Polisher::initialize] loaded sequences
[racon::Polisher::initialize] loaded overlaps
[racon::Polisher::initialize] aligned overlap 181/181
[racon::Polisher::initialize] transformed data into windows
Illegal instruction
```
Would you have any suggestion which type of opcode could cause this error?

I totally understand not using SMID is probably like skipping the main dish :)  Still our first aim is to have software accessible under a generic unknown system :)

**Comment by mmiladi** [comment]
```
(gdb) run sample_reads.fasta sample_overlaps.sam sample_layout.fasta
Starting program: /scratch/bi01/miladim/miniconda3/envs/racon/bin/racon sample_reads.fasta sample_overlaps.sam sample_layout.fasta
[Thread debugging using libthread_db enabled]
[New Thread 0x7ffff5a8f700 (LWP 4080)]
[racon::Polisher::initialize] loaded target sequences
[racon::Polisher::initialize] loaded sequences
[racon::Polisher::initialize] loaded overlaps
[racon::Polisher::initialize] aligned overlap 181/181
[racon::Polisher::initialize] transformed data into windows

Program received signal SIGILL, Illegal instruction.
[Switching to Thread 0x7ffff5a8f700 (LWP 4080)]
0x0000000000425c42 in spoa::SisdAlignmentEngine::initialize(char const*, unsigned int, std::unique_ptr<spoa::Graph, std::default_delete<spoa::Graph> > const&) ()
Missing separate debuginfos, use: debuginfo-install glibc-2.12-1.209.el6_9.2.x86_64

```

**Comment by rvaser** [comment]
Could you find the exact line? I tried to compile locally with the same gcc options and it works smoothly.

**Comment by mmiladi** [comment]
```
(gdb) bt
#0  0x0000000000425c42 in spoa::SisdAlignmentEngine::initialize(char const*, unsigned int, std::unique_ptr<spoa::Graph, std::default_delete<spoa::Graph> > const&) ()
#1  0x00000000004260ce in spoa::SisdAlignmentEngine::align(char const*, unsigned int, std::unique_ptr<spoa::Graph, std::default_delete<spoa::Graph> > const&) ()
#2  0x0000000000426b21 in spoa::SisdAlignmentEngine::align_sequence_with_graph(char const*, unsigned int, std::unique_ptr<spoa::Graph, std::default_delete<spoa::Graph> > const&) ()
#3  0x000000000041eddf in racon::Window::generate_consensus(std::shared_ptr<spoa::AlignmentEngine>) ()
#4  0x000000000040ff7e in std::_Function_handler<bool ()(), std::reference_wrapper<std::_Bind_simple<std::reference_wrapper<std::_Bind<racon::Polisher::polish(std::vector<std::unique_ptr<racon::Sequence, std::default_delete<racon::Sequence> >, std::allocator<std::unique_ptr<racon::Sequence, std::default_delete<racon::Sequence> > > >&, bool)::{lambda(unsigned long)#1} ()(unsigned long)> > ()()> > >::_M_invoke(std::_Any_data const&) ()
#5  0x0000000000414f9a in std::_Function_handler<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()(), std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<bool>, std::__future_base::_Result_base::_Deleter>, bool> >::_M_invoke(std::_Any_data const&) ()
#6  0x000000000041538e in std::__future_base::_State_base::_M_do_set(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()()>&, bool&) ()
#7  0x0000003073e0ce03 in pthread_once () from /lib64/libpthread.so.0
#8  0x000000000040eb2a in std::__future_base::_Task_state<std::_Bind<racon::Polisher::polish(std::vector<std::unique_ptr<racon::Sequence, std::default_delete<racon::Sequence> >, std::allocator<std::unique_ptr<racon::Sequence, std::default_delete<racon::Sequence> > > >&, bool)::{lambda(unsigned long)#1} ()(unsigned long)>, std::allocator<int>, bool ()()>::_M_run() ()
#9  0x000000000040fe05 in std::_Function_handler<void ()(), std::future<std::result_of<racon::Polisher::polish(std::vector<std::unique_ptr<racon::Sequence, std::default_delete<racon::Sequence> >, std::allocator<std::unique_ptr<racon::Sequence, std::default_delete<racon::Sequence> > > >&, bool)::{lambda(unsigned long)#1} ()(unsigned long&)>::type> thread_pool::ThreadPool::submit_task<racon::Polisher::polish(std::vector<std::unique_ptr<racon::Sequence, std::default_delete<racon::Sequence> >, std::allocator<std::unique_ptr<racon::Sequence, std::default_delete<racon::Sequence> > > >&, bool)::{lambda(unsigned long)#1}, unsigned long&>(std::result_of&&, (racon::Polisher::polish(std::vector<std::unique_ptr<racon::Sequence, std::default_delete<racon::Sequence> >, std::allocator<std::unique_ptr<racon::Sequence, std::default_delete<racon::Sequence> > > >&, bool)::{lambda(unsigned long)#1}&&)...)::{lambda()#1}>::_M_invoke(std::_Any_data const&) ()
#10 0x0000000000426ff2 in thread_pool::ThreadPool::worker_thread(thread_pool::ThreadPool*) ()
#11 0x00007ffff7d5ac5c in execute_native_thread_routine_compat () at /opt/conda/conda-bld/compilers_linux-64_1520532893746/work/.build/src/gcc-7.2.0/libstdc++-v3/src/c++11/thread.cc:110
#12 0x0000003073e07aa1 in start_thread () from /lib64/libpthread.so.0
#13 0x00000030736e8bcd in clone () from /lib64/libc.so.6
(gdb) 

```

**Comment by mmiladi** [comment]
Maybe this helps. Sorry I am not so familiar with gdb. 

```
(gdb) f 4
#4  0x000000000040ff7e in std::_Function_handler<bool ()(), std::reference_wrapper<std::_Bind_simple<std::reference_wrapper<std::_Bind<racon::Polisher::polish(std::vector<std::unique_ptr<racon::Sequence, std::default_delete<racon::Sequence> >, std::allocator<std::unique_ptr<racon::Sequence, std::default_delete<racon::Sequence> > > >&, bool)::{lambda(unsigned long)#1} ()(unsigned long)> > ()()> > >::_M_invoke(std::_Any_data const&) ()
(gdb) f 3
#3  0x000000000041eddf in racon::Window::generate_consensus(std::shared_ptr<spoa::AlignmentEngine>) ()
(gdb) f 2
#2  0x0000000000426b21 in spoa::SisdAlignmentEngine::align_sequence_with_graph(char const*, unsigned int, std::unique_ptr<spoa::Graph, std::default_delete<spoa::Graph> > const&) ()
(gdb) f 1
#1  0x00000000004260ce in spoa::SisdAlignmentEngine::align(char const*, unsigned int, std::unique_ptr<spoa::Graph, std::default_delete<spoa::Graph> > const&) ()
(gdb) f 0
#0  0x0000000000425c42 in spoa::SisdAlignmentEngine::initialize(char const*, unsigned int, std::unique_ptr<spoa::Graph, std::default_delete<spoa::Graph> > const&) ()

```

**Comment by rvaser** [comment]
Could you please run gdb/valgrind with `-DCMAKE_BUILD_TYPE=Debug`? I can't find a bug in function initialize (https://github.com/rvaser/spoa/blob/master/src/sisd_alignment_engine.cpp#L64-L119).

**Comment by mmiladi** [comment]
These are available flags from cat /proc/cpuinfo:

`flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm 3dnowext 3dnow constant_tsc rep_good nonstop_tsc extd_apicid amd_dcm pni monitor cx16 popcnt lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt nodeid_msr npt lbrv svm_lock nrip_save pausefilter`

**Comment by mmiladi** [comment]
@rvaser  I think I need to update conda recipe adding `-DCMAKE_BUILD_TYPE=Debug` and wait for anaconda build. What do u think @bgruening ?

**Comment by bgruening** [comment]
We could do this.

**Comment by mmiladi** [comment]
For sure it's not a 'bug', same inputs & binary run on my notebook. Just the opcode is not there. We are now trying to assign the racon jobs to the newer nodes on our server. I'll update you :)

**Comment by rvaser** [comment]
Okay :)

**Comment by mmiladi** [comment]
It worked fine on the newer nodes :) . Now I am updating recipe to only disable `avx2`. This should solve issue for a portion of cases. 
https://github.com/bioconda/bioconda-recipes/pull/9270

**Comment by rvaser** [comment]
Sounds good, most CPUs support sse4 nowadays.

**Comment by rvaser** [comment]
@mmiladi, may I close this issue? :)

**Comment by mmiladi** [comment]
Yes please. Just tested on the ultimate host system and worked very fine :)
FYI: Conda builds have `-DCMAKE_CXX_FLAGS="-mno-avx2 "` 
Thanks & Best,
Milad
